### DISCUSSION TITLE
Open from a link a pdf in a specific page

### DISCUSSION THREAD

**Discussion by science2002** [sumatrapdfreader/sumatrapdf#4067] [Q&A]
I am still with SumatraPDF 3.3 since it works ok in Win10. Thanks for that.
I use to make links of some PDF files for instance in Onenote so to launch them from there. Sometimes I would like to launch the PDFs to a specifc page. But I am not able to do that. 

For instance, let's say I have the following link (in Onenote, manually witten, it should be like that to allow spaces in the file name):
<file:///C:\myFiles\my notebook.pdf>    
This opens in SumatraPDF without problems of course. But when I tried to use the page attribute, I did not find one combination that works. 

One of the most common ones used in other programs is like this:
<file:///C:\myFiles\my notebook.pdf##page=2>    
But does not work. I failed also with other combinations. Can someone kindly suggest if there is one combination that works with SumatraPDF?
Please note that I should not use the following characters in the attribute: " or <>. They are interpreted as the end of the link container  (the beginning being the first <) in Onenote. 
Thanks.

**Comment by GitHubRulesOK** [comment]
Unsure if that is possible without tweaking the registry.
When a file is opened as a URL the Adobe applications use an unsupported method protocol//filename.pdf`#page=2` (note its a single `#`
When using the system it is `app.exe /A page=2` so a different system call altogether.
Browsers expecting URLs will usually pass on the call to the browser PDF plug-in so if you use file://.pdf# it should work in Edge Chrome FireFox

SumatraPDF plug-in did not (AFAIK) support URL fragments (but I could be forgetting) but academic as now depreciated/unsupported

SumatraPDF uses  both `"sumatrapdf.exe" -page 2 "filename.pdf"` AND usually recently `"sumatrapdf.exe" /A page=2 "filename.pdf"` (note in this case the `/A p` does not need double quotes they are considered sequentially.

So when you call a file as a system FTA default only the filetype (.pdf) is expected then opened and if you append .pdf#anything it is not a system acceptable filetype

**Comment by science2002** [comment]
Thanks @GitHubRulesOK for your feedback, and for confirming that when using a link path it is not possible to include also the page number. This is somewhat a pity, because SumatraPDF is so superior in opening files in the correct page (the last one you used) when that PDF was closed recently.

**Comment by GitHubRulesOK** [comment]
I wonder  in cases such as this it may be possible to use a "shim" so if I call the system with a command prefix the desired action in the CMD will be actioned.

So for example, SumatraPDF is my current default (but not on path) so simply calling SumatraPDF does not work but calling a pdf will open at page 1 (or last page closed if remember is active).

However if I have a command **page.cmd** on path  I can use it like this `page 6 "filename.PDF"`

![image](https://github.com/sumatrapdfreader/sumatrapdf/assets/12754913/ef6f20f5-36c7-4057-ad8a-469cfffd14d6)

**page.cmd**
```
@start "SumatraPDF" "C:\Program Files\SumatraPDF\SumatraPDF.exe" -page %1 "%~2"
```

You would need to use the correct path to the exe  and unknown if OneNote will respect path.cmd as a valid system function

As a bonus if you call with a different page SumatraPDF will see the change and reusing the file instance simply goto another page without restart the file.

![image](https://github.com/sumatrapdfreader/sumatrapdf/assets/12754913/00de3d38-1973-4ee7-8714-5c7fb9ccc1fd)

**Comment by science2002** [comment]
Thanks again @GitHubRulesOK

> You would need to use the correct path to the exe and unknown if OneNote will respect path.cmd as a valid system function

Unfortunately it does not work in my several attempts in OneNote. 
If one removes the hyperlink secuirty warning block (from the registry) at best it is able to launch SumatraPDF with this link:
<file:///C:\Program Files\SumatraPDF\SumatraPDF.exe> 
but after the .exe it does not accept anything that works ("Unexpected error" message is the result). Neither it accepts somthing before a standard directory, like C:/etc., or before the code "file:///".
This is at least what I could see in my testing.
But thanks anyway.

**Comment by GitHubRulesOK** [comment]
>it is able to launch SumatraPDF with this link

This is  a weird one as we see later 
So One Not will allow run any single file object without parameters, so there are 2 options.

1) SumatraPDF when opened can offer a selection list of files either 
- 20 ? on home screen but not per page reference
- OR a set of favorites  where specific pages can be selected. 
**2)** OneNote can be cajoled into accepting external LNK files (not easily but doable) so my experience is

You can build a set of shortcut.lnk files with page ,cmd ref so using above method a MyShortCut.Lnk could be 
`"C:\Users\lez\Downloads\demo\New folder (5)\page.cmd" 6 "C:\Users\lez\Downloads\demo\New folder (5)\31.pdf"`

you can test the shortcut to test it works, now the hard part One note will try to override the link contents to just the first file you have to insist manually it is the link you want to open
![image](https://github.com/sumatrapdfreader/sumatrapdf/assets/12754913/48313528-aabe-49cb-889b-be813283f64d)
![image](https://github.com/sumatrapdfreader/sumatrapdf/assets/12754913/817063e6-8707-4c54-98f4-b62a62be1e31)

now it should work !
![image](https://github.com/sumatrapdfreader/sumatrapdf/assets/12754913/93546c37-128c-42e6-9d56-f9b4366652f6)

**Comment by science2002** [comment]
Thanks again @GitHubRulesOK.
Your attempt is an  ingenious method, but a bit cumbersome, if one deals with several hundreds of PDF links, while it could be practical if one just needs few of them.  My case is the former.

Another method could be the following: it requires Autohotkey (AHK), that I already in use (in fact  in my case the manual links <file:/// etc.....> are made by AHK) 
1. Let us adopt the format in use for Acrobat and Edge that is, by adding #page=6 at the end of the fullpath file: it is accepted by Onenote, although with SumatraPDF does - as you confirmed - nothing: it just opens the PDF at the first page. 
2. Build a macro in AHK that monitors the clipboard. When it finds the string "#page=" it manipulates the link in the clipboard so to introduce the path of SumatraPDF,  to rewrite the page format, so to launch the file in the format you have highlighted. 
3.  When one needs that the PDF is launched in the proper page, one just copies the link in the clipboard and AHK does the rest (the advantage of this method is that it works not only in Onenote, but in any other program, even Notepad)
4. If one does not care the page, just clicks the link in Onenote and the PDF opens anyway, although without going in the proper page number.

**Comment by GitHubRulesOK** [comment]
1) I dont think that is allowed only by MS Edge as it converts the command into a browser URL
![image](https://github.com/sumatrapdfreader/sumatrapdf/assets/12754913/7456f827-b978-4e9a-b711-c4f15d8c882b)
2) good luck and perhaps post the method in this thread
3) That is what my one line shim was for,. Simply call page number for any suitable file and SumatraPDF will goto it.
4) you can start with the front page with 20 pinned icons or with the favourites list (just beware the bloating of settings with hundreds of entries) to avoid that you can start with different setting per topic so invoices or comics or bedtime reading

**Comment by science2002** [comment]
In the attempt to test your method @GitHubRulesOK  I found the following:
1. If one inserts in a shortcut lnk "full path of SumatraPDF.exe" -page 2 "full path of the pdf" it works without any further calls, and in that case one could just copy the shourtcut as it is in OneNote (although for security of OneNote double click does not work with the attached shortcut, one can right click and "Open the file" in the contextual menu).
2. However, I notice the following. If the same file was recently opened at let's say pag. 10 ( or whatever) in SumatraPDF,  page 2 attribute does not work. It opens at pag. 10. (I have in SumatraPDF Settings the two "remember" open files settings  flagged, since in general I like the feature).
3. In my case the solution I shall adopt is the simplest one: I shall use the file link <file:///etc. ....> without the page attribute, and add the page number outside the link. So with one click the link opens the pdf, while the page number will be inserted manually in SumatraPDF toolbar.
4.  AHK could do the work, but it needs to constantly run, and monitoring clipboard's content. Someone could not like that. Another solution could be to select the link with page attribute and use a keyboard AHK hotkey to call the script, without monitoring all the time the clipboard. I did not write it, but I know AHK is very good in manipulating text, in particular with these commands: StringReplace, StringGetPos, Stringtrimleft, Stringtrimright.

But let me stress my thanks again, and let me add that SumatraPDF remains by large extend my default reader.

**Reply by GitHubRulesOK** [reply]
I respect all your comments but simply say my test with OneNote "did" respect a shortcut.lnk file ***as long as it was manually inserted*** agreed unwieldy for bulk use and wasting 4 KB on each shortcut as an unwanted double entry.

Depending on VERSION & calling method I have sometimes found you need to send the open page twice (Fist time it may as you say simply trigger a remember, then as it is now open the second repeat call will goto the desired page) You can simulate that by repeating the same command in page,cmd file

**Comment by Eagle20236** [comment]
I use an AutoHotKey script to open a book in Sumatra at a specific page from a list of books.
The book must be listed using the format:
<path to book>\<book title> -page <page number>
eg D:\EBooks\Jacobs - Quantum Mechanics 2nd Ed (2012).pdf -page 20

The section of the script I use is included below. To use the script AutoHotKeys must be running. Copy the book and page number to the clipboard using CTRL-C and then press WIN P. Sumatra will open the book at the required page.

#p:: ;Win-p
{
clipboard =  ; Start off empty to allow ClipWait to detect when the text has arrived.
Send ^c
ClipWait
Book = %clipboard%
Book := RegExReplace(Book, "\.pdf", Replacement := ".pdf""") 	;Insert quote mark after file ext.
Book = "%Book%							;Insert quote mark at front of book path*title
;MsgBox, %Book%
Run,  D:\Program Executables\SumatraPDF\SumatraPDF-3.5.2-64 %Book%	;Open book
}
